Add exercises back to episode 3 (#190)
carpentries-incubator/deep-learning-intro: _episodes/03-monitor-the-model.md

In episode 2 we trained a dense neural network on a *classification task*. For this one hot encoding was used together with a `Categorical Crossentropy` loss function.
This measured how close the distribution of the neural network outputs corresponds to the distribution of the three values in the one hot encoding.
Now we want to work on a *regression task*, thus not predicting a class label (or integer number) for a datapoint. In regression, we like to predict one (and sometimes many) values of a feature. This is typically a floating point number.

> ## Exercise: Architecture of the network
> As we want to design a neural network architecture for a regression task,
> see if you can first come up with the answers to the following questions:
> 1. What must be the dimension of our input layer?

> > 2. How would our output layer look like? What about the activation function? Tip: Remember that the activation function in our previous classification network scaled the outputs between 0 and 1.

Author: The idea is to let the learners think about this a bit before giving them the solution on how to do this directly. But you're right that it's a tricky question, especially the activation function. Would you have suggestions on how to rephrase the question so that we still trigger this thought process, but without making it too difficult?

> > 2. How would our output layer look like? What about the activation function? Tip: Remember that the activation function in our previous classification network scaled the outputs between 0 and 1.

Reviewer: That is a good question. the challenge is, that `softmax` involves quite some math from the point of view of a learner. Here, we could argue, that this math is simply not needed first. This way learners are reassured that they need to consider something else. Then, the concept of the activation function transforming the outputs is important as it has direct consequences onto the variable being regressed.

How about we add a concrete hint:
```suggestion
> 2. We want to to output the prediction of a single number. The output layer of the NN hence cannot be the same as for the classification task earlier. This is because the `softmax` activation being used had a concrete meaning with respect to the class labels which is not needed here. What output layer design would you choose for regression?  
Hint: A layer with `relu` activation, with `sigmoid` activation or no activation at all? 
```

{: .language-python}
![Scatter plot to evaluate training and test set](../fig/03_regression_training_test_comparison.png)

> ## Exercise: Reflecting on our results
> 1. Is the performance of the model as you expected (or better/worse)?
> 2. Is there a noteable difference between training set and test set? And if so, any idea why?
>
> > ## Solution
> >  

> > > While the performance on the train set seems reasonable, the performance on the test set is much worse.

Reviewer: What hint to the cause of this would you give?

Author: We're diving into that in the lesson content that follows of course, but indeed maybe we can give a hint here already.

Reviewer: Ok, so this is just an observation.

Judging from the numbers alone, our neural network preduction would be performing worse than the baseline.

> ## Exercise: Baseline

> > Looking at this baseline: Would you consider this a simple or a hard problem to solve?

Author: Do you think the question itself needs an additional hint?

Judging from the numbers alone, our neural network preduction would be performing worse than the baseline.

> ## Exercise: Baseline
> Looking at this baseline: Would you consider this a simple or a hard problem to solve?
>
> > ## Solution
> >  

> > > This really depends on your definition of hard! The baseline gives a more accurate prediction than just
> > > randomly predicting a number, so the problem is not impossible. However, if no model is going to
> > > perform better than the baseline, the problem may still be very hard.

Reviewer: The last sentence of this solution appears quite confusing to me. Just a suggestion:

```suggestion
> > This really depends on your definition of hard! The baseline gives a more accurate prediction than just
> > randomly predicting a number, so the problem is not impossible to solve with machine learning. However, given the structure of the data and our expectations with respect to quality of prediction, it may remain hard to find a good algorithm which exceeds our baseline by orders of magnitude. 
```

Overfitting is a very common issue and there are many strategies to handle it.
Most similar to classical machine learning might to **reduce the number of parameters**.

> > ## Try to reduce the degree of overfitting by lowering the number of parameters

Reviewer: I personally think that the content of this exercise reads more like teaching content. I would object to make it an exercise. Can you comment on why you took that step?

Reviewer: You could use a different part of the dataset for the exercise.

Author: There is indeed an exercise like that already added in PR #196 . But that's at the very end, as a wrap up exercise, I like that this exercise is about halfway through. My preference is to leave it in as an exercise. I think the exercise can be safely skipped if needed, then there's just the mention of the fact that reducing parameters could work.

> > _________________________________________________________________
> > dense_10 (Dense)             (None, 5)                 55        
> > _________________________________________________________________
> > dense_11 (Dense)             (None, 1)                 6         
> > =================================================================
> > Total params: 961
> > Trainable params: 961
> > Non-trainable params: 0
> >
> > ~~~
> > {:.output}
> >
> > ~~~
> > model.compile(optimizer='adam',
> >               loss='mse',
> >               metrics=[keras.metrics.RootMeanSquaredError()])
> > history = model.fit(X_train, y_train,
> >                     batch_size = 32,
> >                     epochs = 200,
> >                     validation_data=(X_val, y_val), verbose = 2)
> >                     
> > history_df = pd.DataFrame.from_dict(history.history)
> > sns.lineplot(data=history_df[['root_mean_squared_error', 'val_root_mean_squared_error']])
> > plt.xlabel("epochs")
> > plt.ylabel("RMSE")
> > ~~~
> > {:.language-python}
> >
> > ![Output of plotting sample](../fig/03_training_history_3_rmse_smaller_model.png)
> >

> > > There is obviously no single correct solution here. But you will have noticed that the number of nodes can be reduced quiet a bit!

Reviewer: ```suggestion
> > There is no single correct solution here. But you will have noticed that the number of nodes can be reduced quiet a bit!
```
